### PR TITLE
Base pytorch

### PR DESCRIPTION
**Pull request by Kelang-Tian**
Open a new branch to record the construction process of PyTorch. This branch will start from the basic part to facilitate review by reviewers

### CODE REVIEW

File: elasticdl/python/tests/worker_ps_pytorch_test.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+    def tearDown(self):
+        for pserver in self._pservers:
+            pserver.server.stop(0)
+
+    def _seed_torch(self, seed=1029):
+        random.seed(seed)
+        os.environ['PYTHONHASHSEED'] = str(seed)
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed(seed)
```
Comment (skydoorkai) [inline elasticdl/python/tests/worker_ps_pytorch_test.py]: No need to set torch.cuda seed after ` torch.manual_seed`.

File: elasticdl/python/worker/worker_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+            raise RuntimeError(
+                "Only support Allreduce and ParameterServer "
+                "distribution strategy"
+            )
+
+    def _run_model_call_before_training(self, features):
+        """
+        before training: Create variables and report to ps if not created.
+        """
+        _ = self._model.forward(features)
```
Comment (skydoorkai) [inline elasticdl/python/worker/worker_pytorch.py]: No need to call `forward` here for pytorch.

File: elasticdl/python/worker/worker_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+        grads = []
+        grads_tmp = {}
+        for name, parms in self._model.named_parameters():
+            grads_tmp[name] = parms
+        for name in _non_embed_vars_keys:
+            if grads_tmp[name].requires_grad:
+                grads.append(grads_tmp[name].data)
+        return loss, grads
+
+    def forward_process(self, features):
```
Comment (skydoorkai) [inline elasticdl/python/worker/worker_pytorch.py]: Don't put unittest only codes in non-unittest codes.

File: model_zoo/mnist/mnist_subclass_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            return features, tf.cast(r["label"], tf.int32)
+
+    dataset = dataset.map(_parse_data)
+
+    if mode == Mode.TRAINING:
+        dataset = dataset.shuffle(buffer_size=1024)
+    return dataset
+
+
+def eval_metrics_fn():
```
Comment (skydoorkai) [inline model_zoo/mnist/mnist_subclass_pytorch.py:84]: No need to add eval in 1st PR.

Comment (Kelang-Tian) [reply model_zoo/mnist/mnist_subclass_pytorch.py:84]: get it. Just because args need `self.eval`.

File: elasticdl/python/tests/worker_ps_pytorch_test.py (hunk 1 above)
Comment (Kelang-Tian) [reply elasticdl/python/tests/worker_ps_pytorch_test.py]: yes, it is unnecessary.

File: elasticdl/python/worker/worker_pytorch.py (hunk 2 above)
Comment (Kelang-Tian) [reply elasticdl/python/worker/worker_pytorch.py]: maybe like this: `self._model(features)`?

Comment (skydoorkai) [reply elasticdl/python/worker/worker_pytorch.py]: PyTorch nn.Module model is similar to Keras functional model. When we instance a model, all variables are created and initialized. 
For Keras subclass model, variables are lazy-created. Thus, a forward call is needed for subclass model.

Comment (skydoorkai) [reply elasticdl/python/worker/worker_pytorch.py]: PyTorch may need it only when we need to substitute embedding with ElasticDL embedding.

Comment (Kelang-Tian) [reply elasticdl/python/worker/worker_pytorch.py]: Thanks, I figured out the problem.

File: elasticdl/python/tests/worker_trainloop_pytorch_test.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
+# Copyright 2020 The ElasticDL Authors. All rights reserved.
```
Comment (workingloong) [inline elasticdl/python/tests/worker_trainloop_pytorch_test.py]: The filename "worker_pytorch_training_loop_test" may be better.

Comment (Kelang-Tian) [reply elasticdl/python/tests/worker_trainloop_pytorch_test.py]: good suggestion!

File: elasticdl/python/worker/worker_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+    get_dict_from_params_str,
+    get_model_spec,
+    get_non_embedding_trainable_vars,
+    set_callback_parameters,
+)
+from elasticdl.python.common.tensor_utils import EmbeddingTableInfo, Tensor
+from elasticdl.python.common.timing_utils import Timing
+from elasticdl.python.elasticdl.callbacks import SavedModelExporter
+from elasticdl.python.elasticdl.feature_column import feature_column
+from elasticdl.python.elasticdl.layers.embedding import Embedding
```
Comment (workingloong) [inline elasticdl/python/worker/worker_pytorch.py]: Do you use `elasticdl.python.elasticdl.layer`, `elasticdl.python.elasticdl.callbacks`, `elasticdl.python.elasticdl.feature_column` in the pytorch training loop?

Comment (Kelang-Tian) [reply elasticdl/python/worker/worker_pytorch.py]: Actually, these packages are not used. Based training loop doesn't need it.

Comment (Kelang-Tian) [reply elasticdl/python/worker/worker_pytorch.py]: I will remove them later.

File: elasticdl/python/tests/worker_ps_pytorch_test.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        stop_step = 20
+        self._create_worker(1)
+        worker_results = self._worker_train(
+            0, train_db=db, test_db=test_db, stop_step=stop_step
+        )
+        print("show batchs train worker_results[0]\n"
+              "w_loss, acc_meter:", worker_results[0])
+        print("--------finish test_compare_mnist_train!")
+
+    def test_compare_onebatch_train_pytorch(self):
```
Comment (QiJune) [inline elasticdl/python/tests/worker_ps_pytorch_test.py]: We can remove this unit test since it does not do comparing.

File: elasticdl/python/tests/worker_ps_pytorch_test.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+            shape_0 = data_enum[0].shape[0]
+            for i in range(shape_0):
+                dataset_list.append((data_enum[0][i:i + 1, ...], data_enum[1][i, ...]))
+        # Focus: labels.shape [batch_size] , [batch_size, 1] is incorrect
+
+        iterable_dataset = CustomDataset(dataset_list)
+        dataloader = DataLoader(dataset=iterable_dataset, batch_size=batch_size)
+        return dataloader
+
+    def test_compare_mnist_train(self):
```
Comment (QiJune) [inline elasticdl/python/tests/worker_ps_pytorch_test.py]: rename `test_compare_mnist_train ` to `mnist_train`

File: elasticdl/python/tests/worker_pytorch_training_loop_test.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    get_random_batch, )
+
+import sys
+
+sys.path.append('../')
+from worker.worker_pytorch import WorkerPytorch
+import torch
+
+
+class WorkerPSInteractionTest(unittest.TestCase):
```
Comment (QiJune) [inline elasticdl/python/tests/worker_pytorch_training_loop_test.py:36]: It seems that this file is the same with `worker_ps_pytorch_test.py`, we could remove one and keep the other.

**Comment by QiJune** [comment]
It seems that the CI fails due to pre-commit checking, please fix it according to the error message.

File: elasticdl/python/tests/worker_pytorch_training_loop_test.py (hunk 9 above)
Comment (Kelang-Tian) [reply elasticdl/python/tests/worker_pytorch_training_loop_test.py:36]: yes,I rename the file and make the mistake.

File: elasticdl/python/tests/worker_ps_pytorch_test.py (hunk 7 above)
Comment (Kelang-Tian) [reply elasticdl/python/tests/worker_ps_pytorch_test.py]: get it.

File: elasticdl/python/tests/worker_pytorch_training_loop_test.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        self._ports, self._channels, self._pservers = create_pserver(
+            self._model_zoo_path,
+            model_def,
+            grads_to_wait=1,
+            use_async=True,
+            num_ps_pods=num,
+        )
+        self._model_def = model_def
+
+    def _create_worker(self, worker_num):
```
Comment (workingloong) [inline elasticdl/python/tests/worker_pytorch_training_loop_test.py:71]: The function  is  the same as `create_worker` in 
https://github.com/sql-machine-learning/elasticdl/blob/5fe8f4bc01cb0358926066b8d9dd014692e05f26/elasticdl/python/tests/worker_ps_interaction_test.py#L65-L85

You can move both them into `test_utils.py` and import it.

File: elasticdl/python/tests/worker_pytorch_training_loop_test.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+        # Reuse the tf style code to get dataset
+        db, test_db = get_mnist_dataset(self._batch_size)
+
+        stop_step = 20
+        self._create_worker(1)
+        worker_results = self._worker_train(
+            0, train_db=db, test_db=test_db, stop_step=stop_step
+        )
+        print(
```
Comment (workingloong) [inline elasticdl/python/tests/worker_pytorch_training_loop_test.py]: Delete `print` in unit tests.

File: elasticdl/python/tests/worker_pytorch_training_loop_test.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        ]
+        args = parse_worker_args(arguments)
+        worker = WorkerPytorch(args, ps_client=PSClient(self._channels))
+        worker._run_model_call_before_training(images)
+        worker.get_model()
+        w_loss, w_grads = worker.training_process_pytorch(images, labels)
+        worker.report_gradient(w_grads)
+        print("show onebatch train w_loss:", w_loss)
+        print("show onebatch train w_grads[0]:", w_grads[0].shape)
+        print("--------finish test_compare_onebatch_train_pytorch!")
```
Comment (workingloong) [inline elasticdl/python/tests/worker_pytorch_training_loop_test.py]: Delete `print` and you can use assert like `self.assertEqual` to check the test.

File: elasticdl/python/tests/worker_pytorch_training_loop_test.py (hunk 10 above)
Comment (Kelang-Tian) [reply elasticdl/python/tests/worker_pytorch_training_loop_test.py:71]: Thanks. Good suggestion.

File: elasticdl/python/worker/worker_pytorch.py [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+                raise ValueError(
+                    "PS channels are not set up under "
+                    "parameter server strategy"
+                )
+            else:
+                self._model_versions_from_ps = [
+                    -1 for _ in range(self._ps_client.ps_num)
+                ]
+        self._max_minibatch_retry_num = max_minibatch_retry_num
+        self._max_allreduce_retry_num = max_allreduce_retry_num
```
Comment (workingloong) [inline elasticdl/python/worker/worker_pytorch.py:85]: No code uses the variable and you can delete it first.

File: elasticdl/python/worker/worker_pytorch.py [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+            loss=args.loss,
+            optimizer=args.optimizer,
+            eval_metrics_fn=args.eval_metrics_fn,
+            model_params=args.model_params,
+            prediction_outputs_processor=args.prediction_outputs_processor,
+            custom_data_reader=args.custom_data_reader,
+            callbacks=args.callbacks,
+        )
+
+        self._collective_communicator = None
```
Comment (workingloong) [inline elasticdl/python/worker/worker_pytorch.py:122]: No usage and delete it.

File: elasticdl/python/worker/worker_pytorch.py [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+        self._get_model_steps = args.get_model_steps
+        self._opt = self._opt_fn()
+        self._model.optimizer = self._opt
+        self._non_embed_grads = {}
+        self._evaluation_result = {}
+
+    def set_model(self, model_inst):
+        """Set model instance to worker."""
+        self._model = model_inst
+        self._model.float()
```
Comment (workingloong) [inline elasticdl/python/worker/worker_pytorch.py]: What does `self._model.float()` do?

File: elasticdl/python/worker/worker_pytorch.py [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+        self._timing.start_record_time("report_gradient")
+        grads = []
+        for i, v in enumerate(self._non_embed_vars.items()):
+            grad = Tensor(v[0], gradients[i].numpy(), None)
+            grads.append(grad)
+        edl_grads = []
+
+        # TODO: PS in python/go is tf style to get optimizer info,
+        #  need PyTorch style, revise when rewrite pserver
+        learning_rate = K.get_value(self._model.optimizer.lr)
```
Comment (workingloong) [inline elasticdl/python/worker/worker_pytorch.py]: Why do we use the TensorFlow optimizer?

File: elasticdl/python/worker/worker_pytorch.py [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+        # learning_rate = self._model.optimizer.param_groups[0]["lr"]
+
+        accepted, max_version = self._ps_client.push_gradients(
+            grads, edl_grads, learning_rate, self._model_versions_from_ps,
+        )
+        self._timing.end_record_time("report_gradient")
+        return accepted, max_version
+
+    def report_gradient(self, grads):
```
Comment (workingloong) [inline elasticdl/python/worker/worker_pytorch.py]: You can rename `report_gradients_to_ps` to `report_gradient` directly.

File: elasticdl/python/worker/worker_pytorch.py [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+
+        if (
+            self._distribution_strategy
+            == DistributionStrategy.PARAMETER_SERVER
+        ):
+            self._ps_client.partition_dense_parameters(
+                self._non_embed_vars.keys()
+            )
+
+    def _collect_gradients_without_allreduce(self, grads):
```
Comment (workingloong) [inline elasticdl/python/worker/worker_pytorch.py]: `_collect_gradients_without_allreduce` -> `_collect_gradients_with_ps`

File: elasticdl/python/worker/worker_pytorch.py [hunk 19]
```diff
@@ -0,0 +1,400 @@
...
+        for name in _non_embed_vars_keys:
+            if grads_tmp[name].requires_grad:
+                grads.append(grads_tmp[name].data)
+        return loss, grads
+
+    def _run_training_task(self, features, labels):
+        loss, grads = self.training_process_pytorch(features, labels)
+        return (*self._collect_gradients_without_allreduce(grads), loss)
+
+    def _process_minibatch(
```
Comment (workingloong) [inline elasticdl/python/worker/worker_pytorch.py:271]: Can we import the function from `worker.py`?
